src/Listeners/RemoveClaimToken.php

```diff
@@ -15,21 +18,38 @@ class RemoveClaimToken implements ShouldQueue
      */
     public function handle(PlatformBroadcastEvent $event): void
     {
-        BeamClaim::where('token_chain_id', $event->broadcastData['tokenId'])
-            ->whereHas(
-                'collection',
-                fn ($query) => $query->where('collection_chain_id', $event->broadcastData['collectionId'])
-            )
-            ->claimable()
-            ->with('beam')
-            ->get()
-            ->each(function ($claim) {
-                if ($claim->token?->nonFungible) {
-                    $claim->delete();
-                    if ($code = $claim?->beam?->code) {
-                        Cache::decrement(BeamService::key($code));
-                    }
-                }
-            });
+        $collection = Cache::remember(
+            'collection:' . $event->broadcastData['collectionId'],
+            30,
+            fn () => Collection::where('collection_chain_id', $event->broadcastData['collectionId'])->first(['id'])
+        );
+
+        if (!$collection) {
+            return;
+        }
+
+        if (Token::where('collection_id', $collection->id)
+            ->where('token_chain_id', $event->broadcastData['tokenId'])
+            ->first()
+            ?->nonFungible
+        ) {
+            $beamsToDecrement = BeamClaim::where('token_chain_id', $event->broadcastData['tokenId'])
+                ->where('collection_id', $collection->id)
+                ->selectRaw('beam_id, count(*) total')
+                ->claimable()
+                ->groupBy('beam_id')
+                ->pluck('total', 'beam_id');
+
+            BeamClaim::where('token_chain_id', $event->broadcastData['tokenId'])
+                ->where('collection_id', $collection->id)
+                ->claimable()
+                ->delete();
+
+            Beam::whereIn('id', $beamsToDecrement->keys())
+                ->get(['id', 'code'])
+                ->each(fn ($beam) => Cache::decrement(BeamService::key($beam->code), $beamsToDecrement[$beam->id]));
```

Comment:
Can you confirm if there's a `UNIQUE` key associated with `(token_chain_id, collection_id)` on the `beam_claims` table? Otherwise, we can't just assume we want to decrement by `1`. We should move L43-L46 to be deleted as part of `each()` and call `Cache::decrement` on the exact amount of rows deleted.

Reply:
<img width="798" alt="Screenshot 2024-09-03 at 3 51 59 PM" src="https://github.com/user-attachments/assets/68cfe417-1f79-4c83-ac82-427f49666d6c">
I've already resolved  how many will be deleted here, else if we put that in `each` its gonna be more than 1 delete query, were my current implementation is just one query (batch delete)

Reply:
<img width="1100" alt="Screenshot 2024-09-03 at 3 50 28 PM" src="https://github.com/user-attachments/assets/e25f17ae-bf7d-4dc9-9fdb-fbc08c83e15e">
There's no unique key constraint

Reply:
There are two problems that I see:
- The data is not necessarily accurate. In that, between this query and the subsequent query, there's a potential for a race condition. Probability of this happening on a NFT is slim-to-none, but if this is reused for FTs, it's a much higher probability (albeit this code shouldn't ever need to be executed on a FT in the first place — unless it's part of clean-up for burn/destroy);
- If we continue with this approach, we need to cache lock this section and/or job in its entirety so there's a single execution at any given time. I wouldn't cache lock the entire job, as we only really need to prevent a race condition when doing this particular cleanup part. However, if we don't have this, given processing of Beams is handled async, a simple Beam campaign can result in us running this code repetitively and potentially decrementing the `Beam` too much (causing data integrity issues or even underflowing).

Based no the options, I'd probably be in favour of wrapping this part in a `Cache::lock`. Since we're only querying for a list of Beams that have at least one claimable item, this should run once and that'll be the end of it following this cleanup process. Therefore, if there's another job performing this same logic, it'll find nothing to delete and the subsequent queries will not be executed. I would also advise making sure we have at least one `$beamsToDecrement` before making the subsequent calls to the `delete()` and `get()`.